qiskit/transpiler/preset_passmanagers/level0.py

```diff
@@ -261,10 +229,62 @@ def _contains_delay(property_set):
         pm0.append(_unroll)
     if inst_map and inst_map.has_custom_gate():
         pm0.append(PulseGates(inst_map=inst_map))
+
+    # 7. Unify all durations (either SI, or convert to dt if known)
+    # Schedule the circuit only when scheduling_method is supplied
+    # Apply alignment analysis regardless of scheduling for delay validation.
     if scheduling_method:
-        pm0.append(_scheduling)
+        # Do scheduling after unit conversion.
+        scheduler = {
+            "alap": ALAPSchedule,
+            "as_late_as_possible": ALAPSchedule,
+            "asap": ASAPSchedule,
+            "as_soon_as_possible": ASAPSchedule,
+        }
+        pm0.append(TimeUnitConversion(instruction_durations))
+        try:
+            pm0.append(scheduler[scheduling_method](instruction_durations))
+        except KeyError as ex:
+            raise TranspilerError("Invalid scheduling method %s." % scheduling_method) from ex
     elif instruction_durations:
-        pm0.append(_time_unit_setup)
-        pm0.append(_time_unit_conversion, condition=_contains_delay)
-    pm0.append(_alignments)
+        # No scheduling. But do unit conversion for delays.
+        def _contains_delay(property_set):
+            return property_set["contains_delay"]
+
+        pm0.append(ContainsInstruction("delay"))
+        pm0.append(TimeUnitConversion(instruction_durations), condition=_contains_delay)
+    if (
+        timing_constraints.granularity != 1
+        or timing_constraints.min_length != 1
+        or timing_constraints.acquire_alignment != 1
+        or timing_constraints.pulse_alignment != 1
+    ):
+        # Run alignment analysis regardless of scheduling.
+
+        def _require_alignment(property_set):
+            return property_set["reschedule_required"]
+
+        pm0.append(
+            InstructionDurationCheck(
+                acquire_alignment=timing_constraints.acquire_alignment,
+                pulse_alignment=timing_constraints.pulse_alignment,
+            )
+        )
+        pm0.append(
+            ConstrainedReschedule(
+                acquire_alignment=timing_constraints.acquire_alignment,
+                pulse_alignment=timing_constraints.pulse_alignment,
+            ),
+            condition=_require_alignment,
+        )
+        pm0.append(
+            ValidatePulseGates(
+                granularity=timing_constraints.granularity,
+                min_length=timing_constraints.min_length,
+            )
+        )
```

Comment:
Shouldn't `SetIOLatency` be part of preset passmanagers at levels 0-3?

Reply:
It should be but currently the backend data fields are both empty. We don't have any spec for values, i.e. they might be qubit dependent, or a single-value per system, or something else. We can come back to this once we define the format. We can minimize the impact of this unstable API since everything is internal code now.